{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1e8b4751-a921-42ad-938e-9d8b8f820c74",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The prodKeys are ['JUNO_HELIO1DAY_POSITION', 'IA_K0_MFI', 'IA_K0_EPI', 'IA_K0_ENF', 'IA_OR_DEF', 'IA_K0_ICD', 'HELIOS1_COHO1HR_MERGED_MAG_PLASMA', 'HEL1_6SEC_NESSMAG', 'HELIOS1_HELIO1DAY_POSITION', 'HELIOS1_E6_1HOUR_PARTICLE_FLUX']\n",
      "Checking parameter RAD_AU in HAPI record with id JUNO_HELIO1DAY_POSITION\n",
      "The length of data is 1\n",
      "<class 'numpy.float64'>\n",
      "Data was successfully accessed\n",
      "Example data looks like 1.0099999904632568\n",
      "Checking parameter B_abs in HAPI record with id IA_K0_MFI\n",
      "The length of data is 15\n",
      "<class 'numpy.float64'>\n",
      "Data was successfully accessed\n",
      "Example data looks like 1608.0\n",
      "Checking parameter Fe1 in HAPI record with id IA_K0_EPI\n",
      "The length of data is 1\n",
      "<class 'numpy.float64'>\n",
      "Data was successfully accessed\n",
      "Example data looks like -1e+31\n",
      "Checking parameter E1 in HAPI record with id IA_K0_ENF\n",
      "The length of data is 8\n",
      "<class 'numpy.float64'>\n",
      "Data was successfully accessed\n",
      "Example data looks like 8.800000190734863\n",
      "Checking parameter SC_pos_GSE in HAPI record with id IA_OR_DEF\n",
      "The length of data is 3\n",
      "<class 'numpy.ndarray'>\n",
      "Data was successfully accessed\n",
      "Example data looks like [ 0.96600002  0.83399999 -0.61900002]\n",
      "Checking parameter D6_H in HAPI record with id IA_K0_ICD\n",
      "The length of data is 2\n",
      "<class 'numpy.int32'>\n",
      "Data was successfully accessed\n",
      "Example data looks like 1\n",
      "Checking parameter RotationNumber in HAPI record with id HELIOS1_COHO1HR_MERGED_MAG_PLASMA\n",
      "The length of data is 1\n",
      "<class 'numpy.int32'>\n",
      "Data was successfully accessed\n",
      "Example data looks like -2147483648\n",
      "Checking parameter BXSSE in HAPI record with id HEL1_6SEC_NESSMAG\n",
      "The length of data is 300\n",
      "<class 'numpy.float64'>\n",
      "Data was successfully accessed\n",
      "Example data looks like -2.990000009536743\n",
      "Checking parameter RAD_AU in HAPI record with id HELIOS1_HELIO1DAY_POSITION\n",
      "The length of data is 1\n",
      "<class 'numpy.float64'>\n",
      "Data was successfully accessed\n",
      "Example data looks like 0.9800000190734863\n",
      "Checking parameter p_flux in HAPI record with id HELIOS1_E6_1HOUR_PARTICLE_FLUX\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'HAPIError' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mHAPIError\u001b[0m                                 Traceback (most recent call last)",
      "File \u001b[0;32m/srv/conda/envs/notebook/lib/python3.10/site-packages/hapiclient/util.py:311\u001b[0m, in \u001b[0;36murlopen\u001b[0;34m(url)\u001b[0m\n\u001b[1;32m    306\u001b[0m             msg \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mProblem with \u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m+\u001b[39m url \u001b[38;5;241m+\u001b[39m \\\n\u001b[1;32m    307\u001b[0m                   \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m. Server responded with non-200 HTTP status (\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m+\u001b[39m \\\n\u001b[1;32m    308\u001b[0m                   \u001b[38;5;28mstr\u001b[39m(res\u001b[38;5;241m.\u001b[39mstatus) \u001b[38;5;241m+\u001b[39m \\\n\u001b[1;32m    309\u001b[0m                   \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m) but no JSON without HAPI error message in response body.\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m+\u001b[39m c\n\u001b[0;32m--> 311\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m HAPIError\n\u001b[1;32m    313\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m HAPIError:\n",
      "\u001b[0;31mHAPIError\u001b[0m: ",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mHAPIError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn [3], line 13\u001b[0m, in \u001b[0;36mDataChecker\u001b[0;34m(server, dataset, start, stop, parameters)\u001b[0m\n\u001b[1;32m     12\u001b[0m     sleep(\u001b[38;5;241m5.0\u001b[39m)\n\u001b[0;32m---> 13\u001b[0m     data, meta \u001b[38;5;241m=\u001b[39m \u001b[43mhapi\u001b[49m\u001b[43m(\u001b[49m\u001b[43mserver\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdataset\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mparameters\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mstart\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mstop\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     15\u001b[0m \u001b[38;5;66;03m# if error arises\u001b[39;00m\n",
      "File \u001b[0;32m/srv/conda/envs/notebook/lib/python3.10/site-packages/hapiclient/hapi.py:669\u001b[0m, in \u001b[0;36mhapi\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    668\u001b[0m tic0 \u001b[38;5;241m=\u001b[39m time\u001b[38;5;241m.\u001b[39mtime()\n\u001b[0;32m--> 669\u001b[0m \u001b[43murlretrieve\u001b[49m\u001b[43m(\u001b[49m\u001b[43murlbin\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfnamebin\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    670\u001b[0m toc0 \u001b[38;5;241m=\u001b[39m time\u001b[38;5;241m.\u001b[39mtime() \u001b[38;5;241m-\u001b[39m tic0\n",
      "File \u001b[0;32m/srv/conda/envs/notebook/lib/python3.10/site-packages/hapiclient/util.py:368\u001b[0m, in \u001b[0;36murlretrieve\u001b[0;34m(url, fname, check_last_modified, **kwargs)\u001b[0m\n\u001b[1;32m    367\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mopen\u001b[39m(fname, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mwb\u001b[39m\u001b[38;5;124m'\u001b[39m) \u001b[38;5;28;01mas\u001b[39;00m out:\n\u001b[0;32m--> 368\u001b[0m     res \u001b[38;5;241m=\u001b[39m \u001b[43murlopen\u001b[49m\u001b[43m(\u001b[49m\u001b[43murl\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    369\u001b[0m     shutil\u001b[38;5;241m.\u001b[39mcopyfileobj(res, out)\n",
      "File \u001b[0;32m/srv/conda/envs/notebook/lib/python3.10/site-packages/hapiclient/util.py:314\u001b[0m, in \u001b[0;36murlopen\u001b[0;34m(url)\u001b[0m\n\u001b[1;32m    313\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m HAPIError:\n\u001b[0;32m--> 314\u001b[0m     \u001b[43merror\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmsg\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    315\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m urllib3\u001b[38;5;241m.\u001b[39mexceptions\u001b[38;5;241m.\u001b[39mNewConnectionError:\n",
      "File \u001b[0;32m/srv/conda/envs/notebook/lib/python3.10/site-packages/hapiclient/util.py:253\u001b[0m, in \u001b[0;36merror\u001b[0;34m(msg, debug)\u001b[0m\n\u001b[1;32m    251\u001b[0m         sys\u001b[38;5;241m.\u001b[39mexcepthook \u001b[38;5;241m=\u001b[39m exception_handler\n\u001b[0;32m--> 253\u001b[0m \u001b[38;5;28;01mraise\u001b[39;00m HAPIError(msg)\n",
      "\u001b[0;31mHAPIError\u001b[0m: Servlet Name : hapidata;Exception Type : java.lang.NullPointerException;The request URI: /hapi/data;The exception message: null\n",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn [3], line 95\u001b[0m\n\u001b[1;32m     93\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m dataFound:\n\u001b[1;32m     94\u001b[0m     \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mChecking parameter \u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m+\u001b[39m \u001b[38;5;28mstr\u001b[39m(parameters) \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m in HAPI record with id \u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m+\u001b[39m dataset)\n\u001b[0;32m---> 95\u001b[0m     dataFound \u001b[38;5;241m=\u001b[39m \u001b[43mDataChecker\u001b[49m\u001b[43m(\u001b[49m\u001b[43mserver\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdataset\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mstart\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mstop\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mparameters\u001b[49m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[0;32mIn [3], line 16\u001b[0m, in \u001b[0;36mDataChecker\u001b[0;34m(server, dataset, start, stop, parameters)\u001b[0m\n\u001b[1;32m     13\u001b[0m     data, meta \u001b[38;5;241m=\u001b[39m hapi(server, dataset, parameters, start, stop)\n\u001b[1;32m     15\u001b[0m \u001b[38;5;66;03m# if error arises\u001b[39;00m\n\u001b[0;32m---> 16\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[43mHAPIError\u001b[49m:\n\u001b[1;32m     17\u001b[0m     \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mError raised\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     18\u001b[0m     dataFound \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mFalse\u001b[39;00m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'HAPIError' is not defined"
     ]
    }
   ],
   "source": [
    "from datetime import datetime, timedelta\n",
    "from SQLiteFun import execution\n",
    "from hapiclient import hapi\n",
    "from time import sleep\n",
    "\n",
    "# check and see if data can be retrieved from HAPI server\n",
    "def DataChecker(server, dataset, start, stop, parameters, attempts):   \n",
    "    # place in a try/except to wait x seconds and try again if this fails\n",
    "    # if 3 fails in a row -> no data\n",
    "    if attempts < 3:\n",
    "        try:\n",
    "            # Get data\n",
    "            sleep(5.0)\n",
    "            data, meta = hapi(server, dataset, parameters, start, stop)\n",
    "\n",
    "        # if error arises\n",
    "        except:\n",
    "            print(\"Error raised, trying again\")\n",
    "\n",
    "        # if no error arises\n",
    "        else:\n",
    "            dataLen = len(data)\n",
    "            print(\"The length of data is \" + str(dataLen))\n",
    "\n",
    "            # control loop\n",
    "            dataFound = False\n",
    "            # search for data\n",
    "            while not dataFound:\n",
    "                for entry in data:\n",
    "                    print(type(data[0][1]))\n",
    "\n",
    "                    # possible type mismatch when using boolean operator\n",
    "\n",
    "                    # once data is found, end loop\n",
    "                    if str(data[0][1]) != \"\":\n",
    "                        dataFound = True\n",
    "                        print(\"Data was successfully accessed\")\n",
    "                        print(\"Example data looks like \" + str(data[i][1]))\n",
    "                        break\n",
    "        finally:\n",
    "            attempts += 1\n",
    "            return dataFound, attempts\n",
    "    else:\n",
    "        # insert code to input \"HAPI info check passed after 1 attempt. HAPI data check failed after 3 attempts.\"\n",
    "            #     into \"Error\" column in TestResults\n",
    "        # code incorrect, needs to make use of INTERSECT to join the SELECT Spase_ids from MetadataEntries\n",
    "            #     and Update commands since prodKey not a column in TestResults\n",
    "        print(\"No data was found\")\n",
    "        HAPIErrorStmt = f\"\"\" UPDATE TestResults\n",
    "                        SET Errors = 'HAPI info check passed after 1 attempt. HAPI data check failed after 3 attempts.'\n",
    "                        WHERE prodKey = '{dataset}' \"\"\"\n",
    "        Record_id = execution(f\"\"\" SELECT rowNum FROM TestResults WHERE prodKey = '{dataset}';\"\"\")\n",
    "        executionALL(authorStmt)\n",
    "        print(f\"Updated a TestResults entry with the row number {Record_id}\")\n",
    "\n",
    "# retrieve prodKeys associated with HAPI URLs\n",
    "HapiStmt = \"\"\" SELECT prodKey FROM MetadataEntries WHERE url LIKE \"%/hapi\" LIMIT 10\"\"\"\n",
    "prodKeys = execution(HapiStmt)\n",
    "print(\"The prodKeys are \" + str(prodKeys))\n",
    "\n",
    "\n",
    "server = 'https://cdaweb.gsfc.nasa.gov/hapi'\n",
    "#dataset    = 'JUNO_HELIO1DAY_POSITION'\n",
    "#start      = '2011-08-06T00:00:00'\n",
    "#stop       = '2011-08-06T00:30:00'\n",
    "#parameters = 'SE_LAT'\n",
    "#data, meta = hapi(server, dataset, parameters, start, stop)\n",
    "#print(data)\n",
    "\n",
    "# iterate thru prodKeys\n",
    "for prodKey in prodKeys:\n",
    "    # control loop\n",
    "    dataFound = False\n",
    "    # list that holds all parameters for a given server\n",
    "    paramNames =  []\n",
    "\n",
    "    dataset = str(prodKey)\n",
    "\n",
    "    # retrieve all parameters and the start and stop date from the server\n",
    "    meta = hapi(server,dataset)\n",
    "    for k, v in meta.items():\n",
    "        if k == \"parameters\":\n",
    "            for params in v:\n",
    "                for key, value in params.items():\n",
    "                    if key == \"name\":\n",
    "                        paramNames.append(value)\n",
    "        elif k == \"startDate\":\n",
    "            start = v\n",
    "            date, sep, time = start.partition(\"T\")\n",
    "            time = time.replace(\"Z\", \"\")\n",
    "            dt_string = date + \" \" + time\n",
    "            dt_obj = datetime.strptime(dt_string, \"%Y-%m-%d %H:%M:%S\")\n",
    "            stop = dt_obj + timedelta(minutes=30)\n",
    "            stop = str(stop)\n",
    "            stop = stop.replace(\" \", \"T\") + \"Z\"\n",
    "\n",
    "    #print(\"Start time is \" + start)\n",
    "    #print(\"Stop time is \" + stop)\n",
    "    #print(paramNames)\n",
    "    attempts = 0\n",
    "\n",
    "    # iterate thru parameters in a server to see if any data can be pulled\n",
    "    for parameters in paramNames[1:]:\n",
    "        if not dataFound:\n",
    "            print(\"Checking parameter \" + str(parameters) + \" in HAPI record with id \" + dataset)\n",
    "            dataFound, attempts = DataChecker(server, dataset, start, stop, parameters, attempts)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ab063d1-6e9d-40ed-8f5d-f2ed5222ef42",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
